### PR TITLE
Refine the barycoord

### PR DESCRIPTION
**Pull request by AMD-dwang**
1. Separate barycoord from generic builtin variables, because it is
   requried to support interpAt*, so we need to pass more arguments.

2. Move IJCoord operation before the patch.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 1a1040fdf3f5240d0d0ed79fb6155d2e44010683
<details><summary>CTS tests (Failed: 0/289625)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx9</summary><ul>

<li>Passed: 32868/69268 (47.5%)</li>
<li>Failed: 0/69268 (0.0%)</li>
<li>Not Supported: 36400/69268 (52.5%)</li><li>Warnings: 0/69268 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi21</summary><ul>

<li>Passed: 38491/73498 (52.4%)</li>
<li>Failed: 0/73498 (0.0%)</li>
<li>Not Supported: 35007/73498 (47.6%)</li><li>Warnings: 0/73498 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 34348/73361 (46.8%)</li>
<li>Failed: 0/73361 (0.0%)</li>
<li>Not Supported: 39013/73361 (53.2%)</li><li>Warnings: 0/73361 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35388/73498 (48.1%)</li>
<li>Failed: 0/73498 (0.0%)</li>
<li>Not Supported: 38110/73498 (51.9%)</li><li>Warnings: 0/73498 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/builder/InOutBuilder.cpp [hunk 1]
```diff
@@ -767,6 +767,88 @@ Instruction *InOutBuilder::CreateWriteXfbOutput(Value *valueToWrite, bool isBuil
...
+     assert(auxInterpValue != nullptr);
+     auxInterpValue = readBuiltIn(false, BuiltInSamplePosOffset, {}, auxInterpValue, nullptr, "");
+     if (isPersp)
+       auxInterpValue = evalIjOffsetSmooth(auxInterpValue);
+     else
+       auxInterpValue = evalIjOffsetNoPersp(auxInterpValue);
+     break;
+   }
+   default: {
+     // When calling InterpolateAtOffset, auxInterpValue != nullptr
```
Comment (nhaehnle) [inline lgc/builder/InOutBuilder.cpp:817]: Using the `default` in this way is confusing. Better to explicitly have distinct cases for `InterpLocCenter` and `InterpLocExplicit`.

**Review by nhaehnle — COMMENTED** [review]
There is a lot of redundant mess around our barycentric coordinate support, can we please clean that up instead of piling even more confusion onto it?

I see at least the following "built-ins" that appear in LGC IR, handled by PatchInOutImportExport:
```
  case BuiltInInterpPerspSample:
  case BuiltInBaryCoordSmoothSample:
  case BuiltInInterpPerspCenter:
  case BuiltInBaryCoordSmooth:
  case BuiltInInterpPerspCentroid:
  case BuiltInBaryCoordSmoothCentroid:
  case BuiltInInterpPullMode:
  case BuiltInBaryCoordPullModel:
  case BuiltInInterpLinearSample:
  case BuiltInBaryCoordNoPerspSample:
  case BuiltInInterpLinearCenter:
  case BuiltInBaryCoordNoPersp:
  case BuiltInInterpLinearCentroid:
  case BuiltInBaryCoordNoPerspCentroid:
  case BuiltInBaryCoord:
  case BuiltInBaryCoordNoPerspKHR:
```
First, there's the apparent complete redundancy between InterpPerspSample and BaryCoordSmoothSample and so on. That's bad.

Then, `BuiltInBaryCoord[NoPersp]` really does two things: obtain the IJ pixel parameter, and then convert it into the format expected by `VK_KHR_fragment_shader_barycentric`.

What I would suggest is to eliminate the redundancy and split the functionality into orthogonal pieces. Specifically:

1. Remove the LGC BuiltInBaryCoord and BuiltInBaryCoordNoPerspKHR built-ins.
2. Add a CreateBaryCoordNormalization(i, j) builder method which takes as input the IJ coordinates as provided for the HW interpolation view and returns `<3 x float>` IJK coordinates as suitable for the API interpolation view (i.e., taking primitive modes into account).
3. Implement SPIR-V `BuiltInBaryCoord[NoPersp]KHR` in the SPIRV-Reader as "read the appropriate BuiltInBaryCoord{Smooth,NoPersp}*` builtin" + "CreateBaryCoordNormalization".

**Comment by AMD-dwang** [comment]
> 

Thanks your suggests.  I will add a builder method to reorder the barycoord, and remove the `BuiltInBaryCoord` and `BuiltInBaryCoordNoPerspKHR` in PatchInOutImportExport.cpp, but we can't remove them in  lgc built-ins. In the SPIRV-Reader we can't replace them with other built-ins, because there are two way to use `gl_Barycoord` , one is to use directly, the other is as function argument `InterpolateAt*(gl_Barycoord)`, if we replace it, we may get different built-ins. Thanks.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 14a34de78aa1ce35e350e7b9c5247384ec097988
<details><summary>CTS tests (Failed: 0/289625)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx9</summary><ul>

<li>Passed: 32868/69268 (47.5%)</li>
<li>Failed: 0/69268 (0.0%)</li>
<li>Not Supported: 36400/69268 (52.5%)</li><li>Warnings: 0/69268 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi21</summary><ul>

<li>Passed: 38491/73498 (52.4%)</li>
<li>Failed: 0/73498 (0.0%)</li>
<li>Not Supported: 35007/73498 (47.6%)</li><li>Warnings: 0/73498 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 34348/73361 (46.8%)</li>
<li>Failed: 0/73361 (0.0%)</li>
<li>Not Supported: 39013/73361 (53.2%)</li><li>Warnings: 0/73361 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35388/73498 (48.1%)</li>
<li>Failed: 0/73498 (0.0%)</li>
<li>Not Supported: 38110/73498 (51.9%)</li><li>Warnings: 0/73498 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 35c893a7f633b8ba2f3e7a76d13666218535b87e
<details><summary>CTS tests (Failed: 0/289625)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx9</summary><ul>

<li>Passed: 32868/69268 (47.5%)</li>
<li>Failed: 0/69268 (0.0%)</li>
<li>Not Supported: 36400/69268 (52.5%)</li><li>Warnings: 0/69268 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi21</summary><ul>

<li>Passed: 38491/73498 (52.4%)</li>
<li>Failed: 0/73498 (0.0%)</li>
<li>Not Supported: 35007/73498 (47.6%)</li><li>Warnings: 0/73498 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 34348/73361 (46.8%)</li>
<li>Failed: 0/73361 (0.0%)</li>
<li>Not Supported: 39013/73361 (53.2%)</li><li>Warnings: 0/73361 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35388/73498 (48.1%)</li>
<li>Failed: 0/73498 (0.0%)</li>
<li>Not Supported: 38110/73498 (51.9%)</li><li>Warnings: 0/73498 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 300c381e6428bf481d61790a7c330bddf765e74d
<details><summary>CTS tests (Failed: 2/289625)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx9</summary><ul>

<li>Passed: 32866/69268 (47.4%)</li>
<li><details><summary>Failed: 2/69268 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.image.load_store.without_format.buffer.r16g16_sint_minalign
Stack trace: Flake

FAILURE: dEQP-VK.memory.allocation.random.16
Stack trace: Flake
```
</p></details></li>

<li>Not Supported: 36400/69268 (52.5%)</li><li>Warnings: 0/69268 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi21</summary><ul>

<li>Passed: 38491/73498 (52.4%)</li>
<li>Failed: 0/73498 (0.0%)</li>
<li>Not Supported: 35007/73498 (47.6%)</li><li>Warnings: 0/73498 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 34348/73361 (46.8%)</li>
<li>Failed: 0/73361 (0.0%)</li>
<li>Not Supported: 39013/73361 (53.2%)</li><li>Warnings: 0/73361 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 38445/73498 (52.3%)</li>
<li>Failed: 0/73498 (0.0%)</li>
<li>Not Supported: 35053/73498 (47.7%)</li><li>Warnings: 0/73498 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by AMD-dwang** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit 300c381e6428bf481d61790a7c330bddf765e74d
<details><summary>CTS tests (Failed: 0/289625)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx9</summary><ul>

<li>Passed: 32868/69268 (47.5%)</li>
<li>Failed: 0/69268 (0.0%)</li>
<li>Not Supported: 36400/69268 (52.5%)</li><li>Warnings: 0/69268 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi21</summary><ul>

<li>Passed: 38491/73498 (52.4%)</li>
<li>Failed: 0/73498 (0.0%)</li>
<li>Not Supported: 35007/73498 (47.6%)</li><li>Warnings: 0/73498 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 34348/73361 (46.8%)</li>
<li>Failed: 0/73361 (0.0%)</li>
<li>Not Supported: 39013/73361 (53.2%)</li><li>Warnings: 0/73361 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 38445/73498 (52.3%)</li>
<li>Failed: 0/73498 (0.0%)</li>
<li>Not Supported: 35053/73498 (47.7%)</li><li>Warnings: 0/73498 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — DISMISSED** [review]
Thank you!

File: lgc/builder/BuilderRecorder.cpp [hunk 2]
```diff
@@ -168,6 +168,8 @@ StringRef BuilderRecorder::getCallName(Opcode opcode) {
     return "write.generic.output";
   case Opcode::WriteXfbOutput:
     return "write.xfb.output";
+  case Opcode::ReadBaryCoord:
+    return "read.barycoord";
```
Comment (amdrexu) [inline lgc/builder/BuilderRecorder.cpp:172]: barycoord -> bary.coord

File: lgc/builder/BuilderRecorder.cpp [hunk 3]
```diff
@@ -1517,6 +1519,25 @@ Instruction *BuilderRecorder::CreateWriteXfbOutput(Value *valueToWrite, bool isB
                 "");
 }
 
+// Create a read of barycoord input value.
```
Comment (amdrexu) [inline lgc/builder/BuilderRecorder.cpp:1522]: Miss separate line

File: lgc/builder/InOutBuilder.cpp [hunk 4]
```diff
@@ -900,6 +967,93 @@ Value *InOutBuilder::readBuiltIn(bool isOutput, BuiltInKind builtIn, InOutInfo i
...
+  switch (primType) {
+  case PrimitiveType::Point: {
+    // Points
+    barycoord = ConstantVector::get({one, zero, zero});
+  }
+  case PrimitiveType::LineList:
+  case PrimitiveType::LineStrip: {
+    // Lines
+    // The weight of vertex0 is (1 - i - j), the weight of vertex1 is (i + j).
+    auto ly = CreateFAdd(iCoord, jCoord);
```
Comment (amdrexu) [inline lgc/builder/InOutBuilder.cpp:997]: What does ly stand for? Maybe need a more meaningful name.

**Review by amdrexu — APPROVED** [review]

**Comment by AMD-dwang** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit 82be62245f1e95e66836107d054b79284ecf34bc
<details><summary>CTS tests (Failed: 1/216264)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx9</summary><ul>

<li>Passed: 32868/69268 (47.5%)</li>
<li>Failed: 0/69268 (0.0%)</li>
<li>Not Supported: 36400/69268 (52.5%)</li><li>Warnings: 0/69268 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi21</summary><ul>

<li>Passed: 38490/73498 (52.4%)</li>
<li><details><summary>Failed: 1/73498 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.glsl.constants.const_float_from_vec4_vertex
Stack trace: Flake
```
</p></details></li>

<li>Not Supported: 35007/73498 (47.6%)</li><li>Warnings: 0/73498 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35388/73498 (48.1%)</li>
<li>Failed: 0/73498 (0.0%)</li>
<li>Not Supported: 38110/73498 (51.9%)</li><li>Warnings: 0/73498 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by AMD-dwang** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit 82be62245f1e95e66836107d054b79284ecf34bc
<details><summary>CTS tests (Failed: 1/216264)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx9</summary><ul>

<li>Passed: 32867/69268 (47.4%)</li>
<li><details><summary>Failed: 1/69268 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.memory.allocation.random.73
Stack trace: Flake
```
</p></details></li>

<li>Not Supported: 36400/69268 (52.5%)</li><li>Warnings: 0/69268 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi21</summary><ul>

<li>Passed: 38491/73498 (52.4%)</li>
<li>Failed: 0/73498 (0.0%)</li>
<li>Not Supported: 35007/73498 (47.6%)</li><li>Warnings: 0/73498 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35388/73498 (48.1%)</li>
<li>Failed: 0/73498 (0.0%)</li>
<li>Not Supported: 38110/73498 (51.9%)</li><li>Warnings: 0/73498 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by qiaojbao** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit 82be62245f1e95e66836107d054b79284ecf34bc
<details><summary>CTS tests (Failed: 0/216264)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx9</summary><ul>

<li>Passed: 32868/69268 (47.5%)</li>
<li>Failed: 0/69268 (0.0%)</li>
<li>Not Supported: 36400/69268 (52.5%)</li><li>Warnings: 0/69268 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi21</summary><ul>

<li>Passed: 38491/73498 (52.4%)</li>
<li>Failed: 0/73498 (0.0%)</li>
<li>Not Supported: 35007/73498 (47.6%)</li><li>Warnings: 0/73498 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35388/73498 (48.1%)</li>
<li>Failed: 0/73498 (0.0%)</li>
<li>Not Supported: 38110/73498 (51.9%)</li><li>Warnings: 0/73498 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**